nav2_map_server/src/occ_grid_loader.cpp

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) 2008, Willow Garage, Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the Willow Garage, Inc. nor the names of its
+ *       contributors may be used to endorse or promote products derived from
+ *       this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * This file contains helper functions for loading images as maps.
+ *
+ * Author: Brian Gerkey
+ */
+#include "nav2_map_server/occ_grid_loader.hpp"
+
+#include <string>
+#include <vector>
+#include <memory>
+#include <stdexcept>
+#include <chrono>
+
+#include "LinearMath/btQuaternion.h"
+#include "SDL/SDL_image.h"
+
+using namespace std::chrono_literals;
+
+namespace nav2_map_server
+{
+
+const char * OccGridLoader::frame_id_ = "map";
+
+OccGridLoader::OccGridLoader(rclcpp::Node * node)
+: node_(node)
+{
+  std::string mode_str;
+
+  // Get this node's default parameter values, using defaults if not supplied in the YAML file
+  node_->get_parameter_or_set("resolution", resolution_, 0.1);
```

Comment:
Though I understand setting default values for parameters here in the constructor as we've discussed before, I'm slightly concerned we should be careful about which parameters should be set with a default and which should perhaps throw an exception if not set (or logical groups of parameters not set). My rationale is that some default values won't make any sense because they are specific to user configuration. For example with loading maps, an incorrect resolution or origin may throw off the entire map.

Reply:
Sure, let's discuss this topic.